File: clients/julia/test/test_laplace.jl
@testset "Laplace Sampling" verbose = true begin

    BERNOULLI_MODE = "{\"theta\": 0.25}"

    @testset "Data" begin
        out = laplace_sample(bernoulli_model, BERNOULLI_MODE, BERNOULLI_DATA)
        @test 0.22 < mean(get_draws(out, "theta")) < 0.28

        out = laplace_sample(
            bernoulli_model,
            BERNOULLI_MODE,
            joinpath(STAN_FOLDER, "bernoulli", "bernoulli.data.json"),
        )
        @test 0.22 < mean(get_draws(out, "theta")) < 0.28
    end


    @testset "Number of draws" begin
        out =
            laplace_sample(bernoulli_model, BERNOULLI_MODE, BERNOULLI_DATA; num_draws = 234)
        @test size(out.draws, 1) == 234
    end

    @testset "Calculate LP" begin
        out = laplace_sample(
            bernoulli_model,
            BERNOULLI_MODE,
            BERNOULLI_DATA;
            num_draws = 500,
            calculate_lp = true,
        )
        @test sum(isnan.(get_draws(out, "log_p__"))) == 0

        out = laplace_sample(
            bernoulli_model,
            BERNOULLI_MODE,
            BERNOULLI_DATA;
            num_draws = 500,
            calculate_lp = false,
        )
        @test sum(isnan.(get_draws(out, "log_p__"))) == 500
    end

    @testset "Jacobian" begin
        @testset for jacobian in [true, false]
            out_opt =
                optimize(simple_jacobian_model; jacobian = jacobian, seed = UInt32(12345))
            mode_array = out_opt.draws[2:end]
            out = laplace_sample(
                simple_jacobian_model,
                mode_array;
                jacobian = jacobian,
                seed = UInt32(12345),
            )

            optimum = if jacobian
                3.3
            else
                3.0
            end
            @test optimum ≈ mean(get_draws(out, "sigma")) atol = 0.2
        end
    end

    @testset "Save Hessian" begin
        data = "{\"N\": 3}"
        mode = "{\"alpha\": [0.1,0.2,0.3]}"

        out = laplace_sample(gaussian_model, mode, data; save_hessian = true)
        @test size(out.hessian) == (3, 3)
        @test out.hessian ≈ [-1.0 0.0 0.0; 0.0 -1.0 0.0; 0.0 0.0 -1.0] atol = 0.1
    end

    @testset "Seed" begin
        out1 = laplace_sample(
            bernoulli_model,
            BERNOULLI_MODE,
            BERNOULLI_DATA;
            seed = UInt32(123),
        )
        out2 = laplace_sample(
            bernoulli_model,
            BERNOULLI_MODE,
            BERNOULLI_DATA;
            seed = UInt32(123),
        )
        @test out1.draws == out2.draws

        out3 = laplace_sample(
            bernoulli_model,
            BERNOULLI_MODE,
            BERNOULLI_DATA;
            seed = UInt32(456),
        )
        @test out1.draws != out3.draws
    end

    @testset "Bad data" begin
        data1 = "{\"N\": -1}"
        @test_throws "greater than or equal to 0" laplace_sample(
            bernoulli_model,
            BERNOULLI_MODE,
            data1,
        )

        data2 = "{\"N\":1, \"y\": [0,1]}"
        @test_throws "mismatch in dimension" laplace_sample(
            bernoulli_model,
            BERNOULLI_MODE,
            data2,
        )

        @test_throws "Error in JSON parsing" laplace_sample(
            bernoulli_model,
            BERNOULLI_MODE,
            "{'bad'}",
        )
        @test_throws "Could not open data file" laplace_sample(
            bernoulli_model,
            BERNOULLI_MODE,
            "path/not/here.json",
        )
    end

    @testset "Bad mode array" begin
        mode1 = [2.0]
        @test_throws "Bounded variable is 2" laplace_sample(
            bernoulli_model,
            mode1,
            BERNOULLI_DATA,
        )

        mode2 = [1.0]
        @test_throws "incorrect length" laplace_sample(
            gaussian_model,
            mode2,
            "{\"N\": 4 }",
        )
    end

    @testset "Bad mode json" begin
        mode1 = "{\"theta\": 2.0}"
        @test_throws "Bounded variable is 2" laplace_sample(
            bernoulli_model,
            mode1,
            BERNOULLI_DATA,
        )

        mode2 = "{\"theta\": [0.5, 0.5]}"
        @test_throws "mismatch in number" laplace_sample(
            bernoulli_model,
            mode2,
            BERNOULLI_DATA,
        )

        @test_throws "Could not open data file" laplace_sample(
            bernoulli_model,
            "bad/path.json",
            BERNOULLI_DATA,
        )
    end

    @testset "Bad number of draws" begin
        @test_throws "at least 1" laplace_sample(
            bernoulli_model,
            BERNOULLI_MODE,
            BERNOULLI_DATA;
            num_draws = 0,
        )
    end
end

File: clients/python/tests/test_laplace.py
import numpy as np
import pytest

import tinystan
from tests import (
    BERNOULLI_DATA,
    STAN_FOLDER,
    bernoulli_model,
    gaussian_model,
    simple_jacobian_model,
)

BERNOULLI_MODE = {"theta": 0.25}


def test_data(bernoulli_model):
    bernoulli_mode_jac = bernoulli_model.optimize(BERNOULLI_DATA, jacobian=True)
    # data is a string
    out1 = bernoulli_model.laplace_sample(bernoulli_mode_jac, BERNOULLI_DATA)
    assert 0.22 < out1["theta"].mean() < 0.28

    # data stored in a file
    data_file = STAN_FOLDER / "bernoulli" / "bernoulli.data.json"
    out2 = bernoulli_model.laplace_sample(bernoulli_mode_jac, data_file)
    assert 0.22 < out2["theta"].mean() < 0.28


def test_num_draws(bernoulli_model):
    out1 = bernoulli_model.laplace_sample(BERNOULLI_MODE, BERNOULLI_DATA, num_draws=223)
    assert out1["theta"].shape == (223,)


def test_calculate_lp(bernoulli_model):
    out1 = bernoulli_model.laplace_sample(
        BERNOULLI_MODE, BERNOULLI_DATA, num_draws=500, calculate_lp=True
    )
    assert np.sum(np.isnan(out1["log_p__"])) == 0

    out2 = bernoulli_model.laplace_sample(
        BERNOULLI_MODE, BERNOULLI_DATA, num_draws=500, calculate_lp=False
    )
    assert np.sum(np.isnan(out2["log_p__"])) == 500


@pytest.mark.parametrize("jacobian", [True, False], ids=["jacobian", "no_jacobian"])
def test_jacobian(simple_jacobian_model, jacobian):
    mode = simple_jacobian_model.optimize(jacobian=jacobian, seed=12345)

    out = simple_jacobian_model.laplace_sample(mode, jacobian=jacobian, seed=12345)

    if jacobian:
        optimum = 3.3
    else:
        optimum = 3
    assert np.isclose(out["sigma"].mean(), optimum, atol=0.2)


def test_save_hessian(gaussian_model):
    data = {"N": 3}
    mode = {"alpha": [0.1, 0.2, 0.3]}

    out = gaussian_model.laplace_sample(mode, data, save_hessian=True)
    np.testing.assert_almost_equal(
        out.hessian, np.array([[-1, 0, 0], [0, -1, 0], [0, 0, -1]])
    )


def test_seed(bernoulli_model):
    out1 = bernoulli_model.laplace_sample(
        BERNOULLI_MODE,
        BERNOULLI_DATA,
        seed=123,
    )
    out2 = bernoulli_model.laplace_sample(
        BERNOULLI_MODE,
        BERNOULLI_DATA,
        seed=123,
    )

    np.testing.assert_equal(out1["theta"], out2["theta"])

    out3 = bernoulli_model.laplace_sample(
        BERNOULLI_MODE,
        BERNOULLI_DATA,
        seed=456,
    )

    with pytest.raises(AssertionError):
        np.testing.assert_equal(out1["theta"], out3["theta"])


def test_bad_data(bernoulli_model):
    bernoulli_mode_jac = bernoulli_model.optimize(BERNOULLI_DATA, jacobian=True)

    data = {"N": -1}
    with pytest.raises(RuntimeError, match="greater than or equal to 0"):
        bernoulli_model.laplace_sample(bernoulli_mode_jac, data=(data))

    data2 = {"N": 1, "y": [1, 2]}
    with pytest.raises(RuntimeError, match="mismatch in dimension"):
        bernoulli_model.laplace_sample(bernoulli_mode_jac, data=(data2))

    with pytest.raises(RuntimeError, match="Error in JSON parsing"):
        bernoulli_model.laplace_sample(bernoulli_mode_jac, data="{'bad'}")

    with pytest.raises(ValueError, match="Could not open data file"):
        bernoulli_model.laplace_sample(bernoulli_mode_jac, data="path/not/here.json")


def test_bad_mode_stanoutput(bernoulli_model):
    mode1 = bernoulli_model.sample(BERNOULLI_DATA)
    with pytest.raises(ValueError, match="can only be used with"):
        bernoulli_model.laplace_sample(mode1, BERNOULLI_DATA)


def test_bad_mode_array(bernoulli_model):
    mode1 = np.array([2.0])
    with pytest.raises(RuntimeError, match="Bounded variable is 2"):
        bernoulli_model.laplace_sample(mode1, BERNOULLI_DATA)

    mode2 = np.array([])
    with pytest.raises(ValueError, match="incorrect length"):
        bernoulli_model.laplace_sample(mode2, BERNOULLI_DATA)


def test_bad_mode_json(bernoulli_model):
    mode1 = {"theta": 2}  # out of bounds
    with pytest.raises(RuntimeError, match="Bounded variable is 2"):
        bernoulli_model.laplace_sample(mode1, BERNOULLI_DATA)

    mode2 = {"theta": [0.1, 0.1]}
    with pytest.raises(RuntimeError, match="mismatch in number"):
        bernoulli_model.laplace_sample(mode2, BERNOULLI_DATA)

    with pytest.raises(ValueError, match="Could not open data file"):
        bernoulli_model.laplace_sample("bad/path.json", BERNOULLI_DATA)


@pytest.mark.parametrize(
    "arg, value, match",
    [
        ("num_draws", 0, "at least 1"),
    ],
)
def test_bad_argument(bernoulli_model, arg, value, match):
    with pytest.raises(ValueError, match=match):
        bernoulli_model.laplace_sample(BERNOULLI_DATA, **{arg: value})
